recovery/elliptics_recovery/types/merge.py

```diff
@@ -686,6 +683,130 @@ def succeeded(self):
         return self.result
 
 
+class ServerSendRecovery(object):
+    '''
+    Special recovery class that tries to recover keys from backends that
+    should not contain this keys to proper backend via server-send operation.
+    '''
+    def __init__(self, ctx, node, group):
+        self.routes = self._prepare_routes(ctx, group)
+        self.session = elliptics.Session(node)
+        self.session.exceptions_policy = elliptics.exceptions_policy.no_exceptions
+        self.session.set_filter(elliptics.filters.all)
+        self.session.timeout = 60
+        self.session.groups = [group]
+        self.session.trace_id = ctx.trace_id
+        self.ctx = ctx
+
+    def _prepare_routes(self, ctx, group):
+        '''
+        Returns list of triplets (address, backend, [ranges]),
+        where ranges are sorted by their left boundary.
+        '''
+        def sort_ranges(ranges):
+            ranges = sorted(ranges, key=lambda r: r[0])
+            return reduce(lambda x, y: x + y, ranges, tuple())
+
+        group_routes = ctx.routes.filter_by_groups([group])
+        routes = []
+        if ctx.one_node:
+            if ctx.backend_id is not None:
+                ranges = group_routes.get_address_backend_ranges(ctx.address, ctx.backend_id)
+                routes = [(ctx.address, ctx.backend_id, sort_ranges(ranges))]
+            else:
+                for backend_id in group_routes.get_address_backends(ctx.address):
+                    ranges = group_routes.get_address_backend_ranges(ctx.address, backend_id)
+                    routes.append((ctx.address, backend_id, sort_ranges(ranges)))
+        else:
+            for addr, backend_id in group_routes.addresses_with_backends():
+                ranges = group_routes.get_address_backend_ranges(addr, backend_id)
+                routes.append((addr, backend_id, sort_ranges(ranges)))
+
+        log.info("Server-send recovery: group: {0}, num addresses: {1}".format(group, len(routes)))
+        return routes
+
+    def recover(self, keys):
+        '''
+        Tries to recover keys from every backend via server-send. Then it
+        removes keys with older timestamp or invalid checksum.
+        Returns list of keys that was not recovered via server-send.
+        '''
+        log.info("Server-send bucket: num keys: {0}".format(len(keys)))
+
+        def contain(key, ranges):
+            index = bisect(ranges, key)
+            return index % 2 == 1
+
+        responses = dict([(str(k), []) for k in keys]) # key -> [list of responses]
```

Comment:
It is better to use follow form to avoid unnecessary iteration and creating of temporary list:
```python
responses = {str(k): [] for k in keys}
```